### PR TITLE
READY: CLIENTS-347 and Riak command improvements

### PR DESCRIPTION
**Pull request by lukebakken**
Improvements made after working on some of the documentation examples.

More code, but allows chaining of all builder methods.

### CODE REVIEW

File: src/RiakClient/Commands/CommandBuilder{TBuilder,TCommand,TOptions}.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    using System;
+
+    /// <summary>
+    /// Base class for all Riak command builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of the builder. Allows chaining.</typeparam>
+    /// <typeparam name="TCommand">The type of the command.</typeparam>
+    /// <typeparam name="TOptions">The type of the options for this command.</typeparam>
+    public abstract class CommandBuilder<TBuilder, TCommand, TOptions>
+        where TBuilder : CommandBuilder<TBuilder, TCommand, TOptions>
```
Comment (lukebakken) [inline src/RiakClient/Commands/CommandBuilder{TBuilder,TCommand,TOptions}.cs]: This allows chaining by passing the most-derived type as a generic. Each `WithFoo()` method returns the current object cast to the most-derived type:

```csharp
public TBuilder WithFoo(string yomama)
{
    this.yomama = yomama;
    return (TBuilder)this;
}
```

Without this, you can't have `With...()` methods that chain in any order.

File: src/RiakClientExamples/Dev/Using/DataTypes.cs [hunk 2]
```diff
@@ -38,49 +31,106 @@ public sealed class DataTypes : ExampleBase
...
+
+            rslt = client.Execute(updateCmd);
             CheckResult(rslt);
-            Assert.AreEqual(5, rslt.Value);
 
-            rslt = client.DtUpdateCounter(id, -5);
+            response = updateCmd.Response;
+            Assert.AreEqual(5, response.Value);
+
+            updateCmd = new UpdateCounter.Builder(increment: -5, source: builder).Build();
```
Comment (lukebakken) [inline src/RiakClientExamples/Dev/Using/DataTypes.cs]: I :heart: a good old-fashioned copy ctor.

**Comment by alexmoore** [comment]
Looks really nice, great work! +1

**Comment by lukebakken** [comment]
Thanks @alexmoore

**Pull request merged**